packages/libs/src/sdk/services/Solana/programs/PaymentRouterClient/PaymentRouterClient.ts

```diff
@@ -0,0 +1,251 @@
+import { PaymentRouterProgram } from '@audius/spl'
+import {
+  Account,
+  TokenAccountNotFoundError,
+  TokenInvalidMintError,
+  TokenInvalidOwnerError,
+  createAssociatedTokenAccountIdempotentInstruction,
+  createTransferCheckedInstruction,
+  getAccount,
+  getAssociatedTokenAddressSync
+} from '@solana/spl-token'
+import {
+  PublicKey,
+  TransactionInstruction,
+  TransactionMessage,
+  VersionedTransaction
+} from '@solana/web3.js'
+
+import { productionConfig } from '../../../../config/production'
+import { mergeConfigWithDefaults } from '../../../../utils/mergeConfigs'
+import { mintFixedDecimalMap } from '../../../../utils/mintFixedDecimalMap'
+import { parseParams } from '../../../../utils/parseParams'
+import { Prettify } from '../../../../utils/prettify'
+import { Mint } from '../../types'
+import { BaseSolanaProgram } from '../BaseSolanaProgram'
+
+import { getDefaultPaymentRouterClientConfig } from './getDefaultConfig'
+import {
+  CreateMemoInstructionRequest,
+  CreateMemoInstructionSchema,
+  CreatePurchaseContentInstructionsRequest,
+  CreatePurchaseContentInstructionsSchema,
+  CreateRouteInstructionRequest,
+  CreateRouteInstructionSchema,
+  CreateTransferInstructionRequest,
+  CreateTransferInstructionSchema,
+  GetOrCreateProgramTokenAccountRequest,
+  GetOrCreateProgramTokenAccountSchema,
+  PaymentRouterClientConfig
+} from './types'
+
+export class PaymentRouterClient extends BaseSolanaProgram {
+  private readonly programId: PublicKey
+
+  /** The intermediate account where funds are sent to and routed from. */
+  private readonly programAccount: PublicKey
+  private readonly programAccountBumpSeed: number
+
+  private readonly mints: Prettify<Partial<Record<Mint, PublicKey>>>
+
+  private existingTokenAccounts: Prettify<Partial<Record<Mint, Account>>>
+
+  constructor(config: PaymentRouterClientConfig) {
+    const configWithDefaults = mergeConfigWithDefaults(
+      config,
+      getDefaultPaymentRouterClientConfig(productionConfig)
+    )
+    super(configWithDefaults, config.solanaWalletAdapter)
+    this.programId = configWithDefaults.programId
+    const [pda, bump] = PublicKey.findProgramAddressSync(
+      [new TextEncoder().encode('payment_router')],
+      this.programId
+    )
+    this.programAccount = pda
+    this.programAccountBumpSeed = bump
+    this.mints = configWithDefaults.mints
+    this.existingTokenAccounts = {}
+  }
+
+  public async createTransferInstruction(
+    params: CreateTransferInstructionRequest
+  ) {
+    const args = await parseParams(
+      'crateTransferInstruction',
+      CreateTransferInstructionSchema
+    )(params)
+    const mint = this.mints[args.mint]
+    if (!mint) {
+      throw Error('Mint not configured')
+    }
+    const programTokenAccount = await this.getOrCreateProgramTokenAccount({
+      mint: args.mint
+    })
+    const owner = this.wallet.publicKey
+    if (!owner) {
+      throw Error('Connected wallet not found.')
```

Comment:
nit: I'm confused at how the wallet could be set but not have a public key... It feels like the line above is either a potentially unsafe access of `wallet` that isn't typed correctly or `publicKey` shouldn't ever undefined/null.
Any insights?

Reply:
wallet is a `WalletAdapter` object that is not necessarily "connected" at the time. The wallet has a `connect()` async method that can be used to connect. I _think_ this is how things work, but I'm waiting to test it out on staging since we don't have good dev-net testability